url/urlsearchparams.go

```diff
@@ -0,0 +1,403 @@
+}
+
+func newError(r *goja.Runtime, code string, msg string) *goja.Object {
+	o := r.NewTypeError("[" + code + "]: " + msg)
+	o.Set("code", r.ToValue(code))
+	return o
+}
+
+// Currently not supporting the following:
+//   - ctor(iterable): Using function generators
+func createURLSearchParamsConstructor(r *goja.Runtime) goja.Value {
+	f := r.ToValue(func(call goja.ConstructorCall) *goja.Object {
+		u, _ := url.Parse("")
+		v := call.Argument(0)
+		if !goja.IsUndefined(v) {
+			switch v.ExportType() {
+			case reflectTypeString:
+				u = buildParamsFromString(v.String())
+			case reflectTypeObject:
+				u = buildParamsFromObject(r, v)
+			case reflectTypeArray:
+				u = buildParamsFromArray(r, v)
+			case reflectTypeMap:
+				u = buildParamsFromMap(r, v)
+			}
+		}
+
+		sp := parseSearchQuery(u.RawQuery)
+		res := r.ToValue(&nodeURL{url: u, searchParams: sp}).(*goja.Object)
+		res.SetPrototype(call.This.Prototype())
+		return res
+	}).(*goja.Object)
+
+	f.Set("prototype", createURLSearchParamsPrototype(r))
+	return f
+}
+
+// If Parsing results in a path, we move this to the RawQuery
+func buildParamsFromString(s string) *url.URL {
+	u, err := url.Parse(s)
+	if err != nil {
+		return nil
+	}
+
+	if u.Path != "" && u.RawQuery == "" {
+		v, err := url.Parse(fmt.Sprintf("?%s", u.Path))
+		if err != nil {
+			return nil
+		}
+		return v
+	}
+
+	return u
+}
+
+func buildParamsFromObject(r *goja.Runtime, v goja.Value) *url.URL {
+	query := searchParams{}
+
+	o := v.ToObject(r)
+	for _, k := range o.Keys() {
+		val := stringFromValue(r, o.Get(k))
+		query = append(query, searchParam{name: k, value: val})
+	}
+
+	u, _ := url.Parse("")
+	u.RawQuery = query.String()
+	return u
+}
+
+func buildParamsFromArray(r *goja.Runtime, v goja.Value) *url.URL {
+	query := searchParams{}
+
+	o := v.ToObject(r)
+
+	r.ForOf(o, func(val goja.Value) bool {
+		obj := val.ToObject(r)
+		var name, value string
+		i := 0
+		// Use ForOf to determine if the object is iterable
+		r.ForOf(obj, func(val goja.Value) bool {
+			if i == 0 {
+				name = fmt.Sprintf("%v", val)
+				i++
+				return true
+			}
+			if i == 1 {
+				value = fmt.Sprintf("%v", val)
+				i++
+				return true
+			}
+			// Array isn't a tuple
+			panic(newInvalidTupleError(r))
+		})
+
+		// Ensure we have two values
+		if i <= 1 {
+			panic(newInvalidTupleError(r))
+		}
+
+		query = append(query, searchParam{
+			name:  name,
+			value: value,
+		})
+
+		return true
+	})
+
+	u, _ := url.Parse("")
+	u.RawQuery = query.String()
+	return u
+}
+
+func buildParamsFromMap(r *goja.Runtime, v goja.Value) *url.URL {
+	query := searchParams{}
+	o := v.ToObject(r)
+	r.ForOf(o, func(val goja.Value) bool {
+		obj := val.ToObject(r)
+		query = append(query, searchParam{
+			name:  obj.Get("0").String(),
+			value: stringFromValue(r, obj.Get("1")),
+		})
+		return true
+	})
+
+	u, _ := url.Parse("")
+	u.RawQuery = query.String()
+	return u
+}
+
+func stringFromValue(r *goja.Runtime, v goja.Value) string {
```

Comment:
Why do you need this? The standard `.String()` should handle all these cases.

Reply:
Sounds good, I'll remove this.